### ISSUE TITLE
Leaked objects are not reported in the editor, cause crashes on the exported game when the user closes it.

### ISSUE THREAD

**Issue by ericdsw** [godotengine/godot#27103]
**Godot version:** 
v3.1.stable.oficial

**OS/device including version:** 
macOS Mojave

**Issue description:**

I've had a recurrent issue with how Godot reports memory leaks inside the editor vs how the exported game behaves since the 3.0 version. Some examples of these leaks include the following:
- Nodes that were removed from the tree BUT were not freed before the game closes.
- Instanced nodes that were never added to the tree to begin with.

These memory leaks are ignored (or at least, not reported) when closing a game started with the editor's "Play" option, but when running the game from the terminal or with an export, the leak is reported as an error.

Now, I'm aware that in these cases a leak is 100% expected (as stated in the official documentation [here](https://docs.godotengine.org/en/3.0/classes/class_node.html#class-node-remove-child) on how `remove_child` does not actually free the node), but I expected this crash to be reported in the debug project rather than on the final product, which can be a major hassle to fix if the team is not continuously exporting their game's production version.

**Steps to reproduce:**

Note: I've got a `godot` zsh method with the following declaration, just in case it affects anything:
```zsh
godot () {
	/Applications/Godot.app/Contents/MacOS/Godot --path $(pwd) $*
}
```

The easiest way I found to reproduce this was:
1. Run a scene with the following node structure:
```
NodeA (any type)
└──NodeB (any type)
```
2. Add a script that removes `NodeB` from the tree with any logic (can be after one second, on mouse click or in the `_on_ready` method)
3. Close the game once the node is removed.

If executed with the editor's "Play" option, the project closes normally and no crashes occur. However, running it from the command line and following the same steps produces an error with the following output:

```
WARNING: cleanup: ObjectDB Instances still exist!
   At: core/object.cpp:2092.
ERROR: clear: Resources Still in use at Exit!
   At: core/resource.cpp:425.
```

And running it via the terminal with the `-v` flag shows a list of `Leaked instance` and `Orphan StringName` elements, corresponding to the node that was previously removed from the tree with all of its dependencies.

I've noticed that when this is done on an exported project (.exe or .app), the behavior depends on the used version:
- On Godot 3.0, closing the game froze it until killed by the task manager.
- On Godot 3.1 the game closes, but the terminal briefly shows the same error before closing.

**Minimal reproduction project:**

[MemLeakTest.zip](https://github.com/godotengine/godot/files/2972744/MemLeakTest.zip)

In this project I have the following node structure:
```
StartingScene (Node2D)
├── CustomSprite (Sprite)
└── Timer (Timer)
```

After 3 seconds, the timer will call the `_on_Timer_timeout` method inside the script `StartingScene.gd`, which will remove the `CustomSprite` node from the tree. Closing the game after this produces the behavior described above.

Running it via the terminal produces the following output:

```
Using GLES3 video driver
OpenGL debugging not supported!
OpenGL ES 3.0 Renderer: Intel(R) Iris(TM) Graphics 540
CoreAudio: detected 2 channels
CoreAudio: audio buffer frames: 512 calculated latency: 11ms
Unsupported locale 'en_PA', falling back to 'en'.
CORE API HASH: 8676568830230307509
EDITOR API HASH: 66936013638167617
Loading resource: res://default_env.tres
Loaded builtin certs
Loading resource: res://StartingScene.tscn
Loading resource: res://StartingScene.gd
Loading resource: res://CustomSprite.tscn
Loading resource: res://icon.png
Removed sprite
WARNING: cleanup: ObjectDB Instances still exist!
   At: core/object.cpp:2092.
Leaked instance: StreamTexture:1120 - Resource name:  Path: res://icon.png
Leaked instance: Sprite:1128 - Node name: CustomSprite
ERROR: clear: Resources Still in use at Exit!
   At: core/resource.cpp:425.
Orphan StringName: CustomSprite
Orphan StringName: StreamTexture
Orphan StringName: Sprite
StringName: 3 unclaimed string names at exit.
```

**Comment by Zylann** [comment]
When I run a project from the editor with your reproduction steps, I obtain this in the system's console:
```
Running: D:\PROJETS\INFO\GODOT\bin\Godot_v3.1-stable_win64.exe --path D:/PROJETS/INFO/GODOT/TESTS/Godot31/NodeLeak --remote-debug 127.0.0.1:6007 --allow_focus_steal_pid 12028 --position 448,240
OpenGL ES 3.0 Renderer: GeForce GTX 1060 6GB/PCIe/SSE2
WARNING: cleanup: ObjectDB Instances still exist!
     At: core/object.cpp:2092
```
So the bug is always reported, but not directly in the editor.

**Comment by eon-s** [comment]
Related/duplicate of #16815

**Comment by WolfgangSenff** [comment]
Just ran into this, and what's not clear to me is how to fix it. I can no longer run my game from the command line without getting this error, and that in turn prevents me from continuing testing. I don't want to just continue leaking stuff - would rather, if possible, clean up these leaks and then leak again if absolutely necessary - but it's not clear how to do the cleanup manually. Any thoughts?

**Comment by Zylann** [comment]
@WolfgangSenff  If you have no guesses, you can debug a leak by going back in history until the leak doesnt happen (if you have VCS like Git). If this isn't enough to find the culprit change, remove things from your project until it doesn't happen, then you'll know what you removed last is what causes the leak.

**Comment by WolfgangSenff** [comment]
Oh, it did. It leaked a massive amount - 163 objects or so. However, I don't think anything leaks normally - possibly just because of my run through the command line, although I could be mistaken. My problem now is really that it doesn't run through the command line anymore, saying things have leaked. My guess is I'll have to restart my machine in order to recover the memory handles to have it register that I can run again. (For the record, it runs fine through the editor still, just not the command line, but since I'm attempting to get my game to run through a cloud function in Firebase, it will be executed via command line and I can't have it leaking when I do that. :P)

**Comment by ericdsw** [comment]
@WolfgangSenff I've noticed that leaks do happen when running the game from the editor, but they are just not reported in the debugger. When running the editor itself through the command line (or if the editor opens a cmd window like what happens to @Zylann) the error message is shown. Does this happen to you as well?

My main issue with this are not the leaks themselves, but the fact that the editor doesn't report them.

**Comment by akien-mga** [comment]
The reporting of leaked/orphan nodes is now implemented by #28140.

It should not crash the game though, and I don't see evidence of a crash in the OP. If you can still reproduce a crash when orphan nodes are leaked in 3.1 or the master branch, please open a new issue with a stacktrace.
